// openrouteservice/dist/geocode.js
export var GeocodeSearchSource;
(function (GeocodeSearchSource) {
    GeocodeSearchSource["OPENSTREETMAP"] = "osm";
    GeocodeSearchSource["OPENADDRESSES"] = "oa";
    GeocodeSearchSource["WHOS_ON_FIRST"] = "wof";
    GeocodeSearchSource["GEONAMES"] = "gn";
})(GeocodeSearchSource || (GeocodeSearchSource = {}));
export var GeocodeSearchLayer;
(function (GeocodeSearchLayer) {
    GeocodeSearchLayer["ADDRESS"] = "address";
    GeocodeSearchLayer["VENUE"] = "venue";
    GeocodeSearchLayer["NEIGHBORHOOD"] = "neighbourhood";
    GeocodeSearchLayer["LOCALITY"] = "locality";
    GeocodeSearchLayer["BOROUGH"] = "borough";
    GeocodeSearchLayer["LOCAL_ADMINISTRATION"] = "localadmin";
    GeocodeSearchLayer["COUNTY"] = "county";
    GeocodeSearchLayer["MACRO_COUNTY"] = "macrocounty";
    GeocodeSearchLayer["REGION"] = "region";
    GeocodeSearchLayer["MACRO_REGION"] = "macroregion";
    GeocodeSearchLayer["COUNTRY"] = "country";
    GeocodeSearchLayer["COARSE"] = "coarse";
})(GeocodeSearchLayer || (GeocodeSearchLayer = {}));
